src/menhir_parser.mly

```diff
@@ -100,13 +100,31 @@ let translate_old_rule : old_p_rule -> p_rule = fun r ->
     match h.elt with
     | P_Iden({elt = ([],x); _}, _) when is_pat_var env x ->
        let lts = List.map (fun (p,t) -> p,build env t) lts in
-       let n =
-         try Hashtbl.find arity x with Not_found ->
-           assert false (* Unreachable. *)
-       in
-       let (lts1, lts2) = List.cut lts n in
-       let ts1 = Array.of_list (List.map snd lts1) in
-       add_args (Pos.make t.pos (P_Patt(Pos.make h.pos x, ts1))) lts2
+       let max = try Hashtbl.find arity x with Not_found -> assert false in
+       let k = List.length lts in
+       let nb_exp = if k >= max then 0 else max - k in
+       let p = t.pos in
+       if nb_exp = 0 then
+         let (lts1, lts2) = List.cut lts max in
+         let ts1 = Array.of_list (List.map snd lts1) in
+         let patt = Pos.make p (P_Patt(Pos.make h.pos x, ts1)) in
+         add_args patt lts2
+       else (* We eta-expense. *)
+         let ts = Array.of_list (List.map snd lts) in
+         let prefix = "var" in
+         let arg i =
+           if i < k then ts.(i)
+           else
+             let qid = Pos.make p ([], prefix ^ string_of_int (i-k)) in
+             Pos.make p (P_Iden (qid, false))
+         in
+         let args = Array.init max arg in
+         let patt = Pos.make p (P_Patt(Pos.make h.pos x, args)) in
+         let rec vars i =
+           if i >= nb_exp then []
+           else Some (Pos.make p (prefix ^ string_of_int i)) :: vars (i+1)
+         in
+         Pos.make t.pos (P_Abst([vars 0, None, false], patt))
```

Comment:
I'll review that part later, I need to open the file for that (I don't have time now).

Reply:
I can already say that there can be some unintended variable capture here: you should use special name, not accessible to the user for the bound variables here. For example you could take `"@x"` for the prefix. Or at least check for collisions when following your naming convention (skipping the indices that are already used).

Reply:
You are right. I am not in favor of using "@..." because this is not parsable.

Reply:
Should be fixed in 376bf6e